pkg/operator/staticpodcontroller/installer_controller.go

```diff
@@ -0,0 +1,348 @@
+	ret := currKubeletState.DeepCopy()
+	installerPod, err := c.kubeClient.CoreV1().Pods(c.targetNamespace).Get(getInstallerPodName(currKubeletState.TargetDeploymentID, currKubeletState.NodeName), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		ret.LastFailedDeploymentID = currKubeletState.TargetDeploymentID
+		ret.TargetDeploymentID = currKubeletState.CurrentDeploymentID
+		ret.Errors = []string{err.Error()}
+		return ret, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	switch installerPod.Status.Phase {
+	case corev1.PodSucceeded:
+		ret.CurrentDeploymentID = currKubeletState.TargetDeploymentID
+		ret.TargetDeploymentID = 0
+		ret.LastFailedDeploymentID = 0
+		ret.Errors = nil
+	case corev1.PodFailed:
+		ret.LastFailedDeploymentID = currKubeletState.TargetDeploymentID
+		ret.TargetDeploymentID = 0
+
+		errors := []string{}
+		for _, containerStatus := range installerPod.Status.ContainerStatuses {
+			if containerStatus.State.Terminated != nil && len(containerStatus.State.Terminated.Message) > 0 {
+				errors = append(errors, fmt.Sprintf("%q: %s", containerStatus.Name, containerStatus.State.Terminated.Message))
+			}
+		}
+		if len(errors) == 0 {
+			errors = append(errors, "no detailed termination message, see `oc get -n %q pods/%q -oyaml`", installerPod.Namespace, installerPod.Name)
+		}
+		ret.Errors = errors
+	}
+
+	return ret, nil
+}
+
+// getDeploymentIDToStart returns the deploymentID we need to start or zero if none
+func (c *InstallerController) getDeploymentIDToStart(currKubeletState, prevKubeletState *KubeletState, operatorStatus *StaticPodConfigStatus) int32 {
+	if prevKubeletState == nil {
+		currentAtLatest := currKubeletState.CurrentDeploymentID == operatorStatus.LatestDeploymentID
+		failedAtLatest := currKubeletState.LastFailedDeploymentID == operatorStatus.LatestDeploymentID
+		if !currentAtLatest && !failedAtLatest {
+			return operatorStatus.LatestDeploymentID
+		}
+		return 0
+	}
+
+	prevInTransition := prevKubeletState.CurrentDeploymentID != prevKubeletState.TargetDeploymentID
+	if prevInTransition {
+		return 0
+	}
+
+	prevAhead := currKubeletState.CurrentDeploymentID > currKubeletState.CurrentDeploymentID
+	failedAtPrev := currKubeletState.LastFailedDeploymentID == prevKubeletState.CurrentDeploymentID
+	if prevAhead && !failedAtPrev {
+		return currKubeletState.CurrentDeploymentID
+	}
+
+	return 0
+}
+
+func getInstallerPodName(deploymentID int32, nodeName string) string {
+	return fmt.Sprintf("installer-%d-%s", deploymentID, nodeName)
+}
+
+// createInstallerPod creates the installer pod with the secrets required to
+func (c *InstallerController) createInstallerPod(currKubeletState *KubeletState, operatorSpec *operatorv1alpha1.OperatorSpec, deploymentID int32) (*KubeletState, error) {
+	required := resourceread.ReadPodV1OrDie(v311_00_assets.MustAsset("v3.11.0/kube-apiserver/installer-pod.yaml"))
+	switch corev1.PullPolicy(operatorSpec.ImagePullPolicy) {
```

Comment:
Why we care about this validation here? Or if we do validate why not other fields too?

Reply:
> Why we care about this validation here? Or if we do validate why not other fields too?

Shrugs. @mfojtik's seems reasonable.